<a href="https://colab.research.google.com/github/cristiandarioortegayubro/BA/blob/main/dc_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![logo](https://www.python.org/static/community_logos/python-powered-h-100x130.png)

# **Ejercicio, un poquito de todo...**

Nos encontramos iniciando un nuevo proyecto, nuestra empresa es una pequeña inmobiliaria que quiere fusionar el potencial del manejo de datos con la venta/alquiler de propiedades. Al ser una pequeña empresa el ***departamento de datos*** esta compuesto por una persona: nosotros. 

Esto quiere decir que vamos a absorber los roles de:

- Ingeniero de Datos, y 
- Analista de Datos.


## **Ingesta de datos de propiedades**

***El dataset central de este proyecto de propiedades de Argentina, tiene la siguientes columnas:***

- **id** - Identificador del aviso. No es único: si el aviso es actualizado por la inmobiliaria (nueva versión del aviso) se crea un nuevo registro con la misma id pero distintas fechas: de alta y de baja.
- **ad_type** - Tipo de aviso (Propiedad, Desarrollo/Proyecto).
- **start_date** - Fecha de alta del aviso.
- **end_date** - Fecha de baja del aviso.
- **created_on** - Fecha de alta de la primera versión del aviso.
- **lat** - Latitud.
- **lon** - Longitud.
- **l1** - Nivel administrativo 1: país.
- **l2** - Nivel administrativo 2: usualmente provincia.
- **l3** - Nivel administrativo 3: usualmente ciudad.
- **l4** - Nivel administrativo 4: usualmente barrio.
- **l5** - sin informacion
- **l6** - sin informacion
- **rooms** - Cantidad de ambientes
- **bedrooms** - Cantidad de dormitorios
- **bathrooms** - Cantidad de baños.
- **surface_total** - Superficie total en m².
- **surface_covered** - Superficie cubierta en m².
- **price** - Precio publicado en el anuncio.
- **currency** - Moneda del precio publicado.
- **price_period** - Periodo del precio (Diario, Semanal, Mensual)
- **title** - Título del anuncio.
- **description** - Descripción del anuncio.
- **property_type**: - Caracteristicas de la propiedad
- **operation_type**: - Caracteristicas de la operacion ['Venta', 'Alquiler', 'Alquiler temporal']

## **Habilitando bibliotecas**

***Habilitar las bibliotecas necesarias para generar un dataframe, hacer limpieza de datos, y generar gráficos.***

### Para analisis de datos

In [ ]:
import numpy as np
import pandas as pd

### Para gráficos

In [ ]:
import plotly.express as px

## **Análisis Exploratorio de los Datos**

***Leer el archivo  ```.csv``` con Pandas y guardarlo en una variable llamada ```prop```***

In [ ]:
prop = pd.read_csv("/content/drive/MyDrive/Notebooks/Formacion/propiedades.csv")

***Observar los primeros 8 registros***

In [ ]:
prop.head(8)

id    ad_type  start_date    end_date  created_on  \
0  DyVXfkpKygVBKuUk5olH+A==  Propiedad  2020-08-22  2020-09-03  2020-08-22   
1  9naojilaMecJN4jlQiTkGg==  Propiedad  2020-08-22  2020-09-04  2020-08-22   
2  tlCPRJPjoDEUzuuCelemAQ==  Propiedad  2020-08-22  2020-08-31  2020-08-22   
3  Zw3b91glQUO3HNrM5fPYlQ==  Propiedad  2020-08-22  2020-09-04  2020-08-22   
4  bsU81gm9JEgtZCbTYgvykg==  Propiedad  2020-08-22  2020-09-04  2020-08-22   
5  3Cips/GEPuJqktYycSALCQ==  Propiedad  2020-08-22  2020-09-04  2020-08-22   
6  6TO5dWwmivuJ67zkztUzCA==  Propiedad  2020-08-22  2020-09-04  2020-08-22   
7  VlE71iojfIFoH4piJfFQ/g==  Propiedad  2020-08-22  2020-09-04  2020-08-22   

         lat        lon         l1                            l2  \
0 -34.407468 -58.957367  Argentina      Bs.As. G.B.A. Zona Norte   
1 -37.996039 -57.542509  Argentina  Buenos Aires Costa Atlántica   
2 -31.380187 -58.009182  Argentina                    Entre Ríos   
3 -27.494106 -55.123455  Argentina                      Misiones   
4 -32.948856 -60.630464  Argentina                      Santa Fe   
5 -32.884278 -60.710901  Argentina                      Santa Fe   
6 -34.604195 -58.559827  Argentina      Bs.As. G.B.A. Zona Oeste   
7 -32.948834 -60.630169  Argentina                      Santa Fe   

                l3  ... bathrooms surface_total  surface_covered  price  \
0            Pilar  ...       NaN      133139.0              NaN    NaN   
1    Mar del Plata  ...       NaN         687.0            687.0    NaN   
2        Concordia  ...       1.0          80.0             80.0    NaN   
3            Oberá  ...       1.0           NaN              NaN    NaN   
4          Rosario  ...       1.0          76.0             66.0    NaN   
5          Rosario  ...       1.0          39.0             33.0    NaN   
6  Tres de Febrero  ...       1.0           NaN             42.0    NaN   
7          Rosario  ...       2.0         130.0            108.0    NaN   

   currency  price_period                                              title  \
0       NaN       Mensual  VENTA - Lote Industrial 130.000m2 - Parque Ind...   
1       NaN       Mensual                               Casa - Santa Cecilia   
2       NaN       Mensual                                   Casa - Concordia   
3       NaN       Mensual                                    Terreno - Obera   
4       NaN       Mensual                             Departamento - Rosario   
5       NaN       Mensual                               9 DE JULIO 535 06 01   
6       NaN       Mensual                             Departamento - Caseros   
7       NaN       Mensual                             Departamento - Rosario   

                                         description  property_type  \
0  Lote (Nro.50) mide 133.139 m2\n<br>Valor: U$S ...           Lote   
1  Conjunto edilicio de casa y departamentos en b...           Otro   
2  Casa de un dormitorio, cocina, baño, living co...           Casa   
3  IMPORTANTE PROPIEDAD EN VENTA EN OBERA MISIONE...           Lote   
4  PORTAL DE LOS MARINOS<br>Es un complejo de viv...   Departamento   
5  CARACTERÍSTICAS<br>Ingreso living comedor con ...   Departamento   
6  Departamento 2 ambientes con Lavadero en Nicar...   Departamento   
7  PORTAL DE LOS MARINOS<br>Es un complejo de viv...   Departamento   

  operation_type  
0          Venta  
1          Venta  
2       Alquiler  
3          Venta  
4          Venta  
5       Alquiler  
6       Alquiler  
7          Venta  

[8 rows x 25 columns]

***Observar los últimos 5 registros***

In [ ]:
prop.tail()

id    ad_type  start_date    end_date  \
999995  jMYhwydKqUje8HCx7RzwJQ==  Propiedad  2020-12-04  9999-12-31   
999996  FsP7ornzqnK1CvUL8quQ9Q==  Propiedad  2020-12-04  9999-12-31   
999997  fJLVUIziPl31ACoaE/fj1g==  Propiedad  2020-12-04  9999-12-31   
999998  7mW3bUSJFA0RBOj2xtBi+w==  Propiedad  2020-12-04  9999-12-31   
999999  rEynGyGrrsxNeG0ib5vUBA==  Propiedad  2020-12-04  2020-12-21   

        created_on        lat        lon         l1                        l2  \
999995  2020-12-04 -32.958863 -60.665633  Argentina                  Santa Fe   
999996  2020-12-04 -34.596587 -58.376893  Argentina           Capital Federal   
999997  2020-12-04 -34.576445 -58.432603  Argentina           Capital Federal   
999998  2020-12-04 -32.866636 -68.841144  Argentina                   Mendoza   
999999  2020-12-04 -34.447524 -58.631552  Argentina  Bs.As. G.B.A. Zona Norte   

             l3  ... bathrooms surface_total  surface_covered     price  \
999995  Rosario  ...       7.0         400.0            300.0  400000.0   
999996   Retiro  ...       5.0         465.0            465.0    3700.0   
999997  Palermo  ...       2.0         615.0            425.0  570000.0   
999998      NaN  ...      20.0         450.0            450.0  800000.0   
999999    Tigre  ...       4.0         350.0            300.0    7000.0   

        currency  price_period  \
999995       USD       Mensual   
999996       USD       Mensual   
999997       USD       Mensual   
999998       USD       Mensual   
999999       USD       Mensual   

                                                    title  \
999995  Se vende imponente casa a minutos del centro d...   
999996                        Edificio Comercial - Retiro   
999997  Edificio Comercial  en excelente zona de Paler...   
999998                       Edificio Comercial - Capital   
999999  INCREÍBLE ECASA A LA LAGUNA DESDE DIC a FEBRER...   

                                              description    property_type  \
999995  Esta excelente propiedad se encuentra ubicada ...             Casa   
999996  GRAN EDIFICIO PETIT HOTEL ESTILO FRANCÉS  CON ...  Local comercial   
999997  Edificio Comercial CON ENTRADA INDEPENDIENTE S...  Local comercial   
999998  Local 450m² con Cloaca en 25 de mayo, Argentin...  Local comercial   
999999  ESPECTACULAR CASA A LA LAGUNA  -BARRIO SANTA B...             Casa   

           operation_type  
999995              Venta  
999996           Alquiler  
999997              Venta  
999998              Venta  
999999  Alquiler temporal  

[5 rows x 25 columns]

***Imprimir la cantidad de filas y columnas del dataframe***

In [ ]:
prop.shape
print(f"El dataset contiene {prop.shape[0]} filas y {prop.shape[1]} columnas")

El dataset contiene 1000000 filas y 25 columnas


**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $1000000$ filas y $25$ columnas...***

**Usar info**

In [ ]:
prop.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 25 columns):
 #   Column           Non-Null Count    Dtype  
---  ------           --------------    -----  
 0   id               1000000 non-null  object 
 1   ad_type          1000000 non-null  object 
 2   start_date       1000000 non-null  object 
 3   end_date         1000000 non-null  object 
 4   created_on       1000000 non-null  object 
 5   lat              894233 non-null   float64
 6   lon              894191 non-null   float64
 7   l1               1000000 non-null  object 
 8   l2               1000000 non-null  object 
 9   l3               965273 non-null   object 
 10  l4               306162 non-null   object 
 11  l5               5530 non-null     object 
 12  l6               0 non-null        float64
 13  rooms            714179 non-null   float64
 14  bedrooms         649933 non-null   float64
 15  bathrooms        765122 non-null   float64
 16  surface_total    47

***Usar ```describe()``` para variables cuantitativas y mostrar el resumen con dos decimales y transpuesto (transposicion)...***

In [ ]:
round(prop.describe(),2).T

count       mean         std     min       25%       50%  \
lat              894233.0     -34.38        3.04  -54.84    -34.72    -34.59   
lon              894191.0     -59.49        2.80 -122.54    -58.88    -58.49   
l6                    0.0        NaN         NaN     NaN       NaN       NaN   
rooms            714179.0       2.92        1.74    1.00      2.00      3.00   
bedrooms         649933.0       2.10        2.08  -16.00      1.00      2.00   
bathrooms        765122.0       1.71        1.09    1.00      1.00      1.00   
surface_total    477831.0     513.76     4297.24 -136.00     50.00     95.00   
surface_covered  487756.0   11656.22  4443191.26 -130.00     45.00     78.00   
price            958243.0  346094.02  5713689.88    0.00  38000.00  97000.00   

                       75%           max  
lat                 -34.42  4.963000e+01  
lon                 -58.40  1.800000e+02  
l6                     NaN           NaN  
rooms                 4.00  4.000000e+01  
bedrooms              3.00  9.000000e+02  
bathrooms             2.00  2.000000e+01  
surface_total       250.00  2.000000e+05  
surface_covered     168.00  2.147484e+09  
price            210000.00  3.100000e+09

***Usar ```describe()``` para variables cualitativas y mostrar resumen transpuesto...***

In [ ]:
prop.describe(include=[object]).T

count   unique  \
id              1000000  1000000   
ad_type         1000000        1   
start_date      1000000      362   
end_date        1000000      450   
created_on      1000000      362   
l1              1000000        4   
l2              1000000       43   
l3               965273     1367   
l4               306162     1070   
l5                 5530       21   
currency         955491        5   
price_period     429870        3   
title            999999   548833   
description      999958   817449   
property_type   1000000       10   
operation_type  1000000        3   

                                                           top     freq  
id                                    DyVXfkpKygVBKuUk5olH+A==        1  
ad_type                                              Propiedad  1000000  
start_date                                          2020-10-02    25159  
end_date                                            9999-12-31   208960  
created_on                                          2020-10-02    25159  
l1                                                   Argentina   983115  
l2                                             Capital Federal   265125  
l3                                                     Rosario    61746  
l4                                                    Nordelta    19479  
l5                                            BarrioPortezuelo      735  
currency                                                   USD   738157  
price_period                                           Mensual   429842  
title                                               casa-venta     3338  
description     Departamento en Alquiler de 1 dorm. en Cordoba     1980  
property_type                                     Departamento   457907  
operation_type                                           Venta   750831

***Ver si existen valores nulos***

In [ ]:
prop.isnull().sum().sort_values(ascending = False)

l6                 1000000
l5                  994470
l4                  693838
price_period        570130
surface_total       522169
surface_covered     512244
bedrooms            350067
rooms               285821
bathrooms           234878
lon                 105809
lat                 105767
currency             44509
price                41757
l3                   34727
description             42
title                    1
property_type            0
id                       0
ad_type                  0
l2                       0
l1                       0
created_on               0
end_date                 0
start_date               0
operation_type           0
dtype: int64

***Ver si existen valores duplicados***

In [ ]:
prop.duplicated().sum()

0

**PUNTO DE REVISIÓN**

***Antes de empezar con la limpieza de los datos el dataframe tiene $1000000$ filas y $25$ columnas...***

## **Limpieza de los Datos**

***Ver si los nombres de las columnas tienen algun espacio oculto que nos dificulte al tratar de invocarlas, si encontras espacios ocultos, entonces se debe eliminar esos espacios ocultos.***

In [ ]:
prop.columns

Index(['id', 'ad_type', 'start_date', 'end_date', 'created_on', 'lat', 'lon',
       'l1', 'l2', 'l3', 'l4', 'l5', 'l6', 'rooms', 'bedrooms', 'bathrooms',
       'surface_total', 'surface_covered', 'price', 'currency', 'price_period',
       'title', 'description', 'property_type', 'operation_type'],
      dtype='object')

***Ver los nulos que hay en el dataframe, en forma de porcentaje para cada columna, ordenado del mayor al menor...***

In [ ]:
round((prop.isna().sum()*100/prop.shape[0]),2).sort_values(ascending = False)

l6                 100.00
l5                  99.45
l4                  69.38
price_period        57.01
surface_total       52.22
surface_covered     51.22
bedrooms            35.01
rooms               28.58
bathrooms           23.49
lat                 10.58
lon                 10.58
currency             4.45
price                4.18
l3                   3.47
title                0.00
description          0.00
property_type        0.00
id                   0.00
ad_type              0.00
l2                   0.00
l1                   0.00
created_on           0.00
end_date             0.00
start_date           0.00
operation_type       0.00
dtype: float64

***Eliminar columnas que tengan mas del $30$% de nulos (nans), luego de este paso deberian quedar $18$ columnas si lo hiciste bien...***

In [ ]:
prop.shape

(1000000, 25)

In [ ]:
prop.dropna(axis = 1, inplace = True, thresh = (1000000 * 0.7))

In [ ]:
prop.shape

(1000000, 18)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $1000000$ filas y $18$ columnas...***

In [ ]:
round((prop.isna().sum()*100/prop.shape[0]),2).sort_values(ascending = False)

rooms             28.58
bathrooms         23.49
lat               10.58
lon               10.58
currency           4.45
price              4.18
l3                 3.47
property_type      0.00
description        0.00
title              0.00
id                 0.00
ad_type            0.00
l2                 0.00
l1                 0.00
created_on         0.00
end_date           0.00
start_date         0.00
operation_type     0.00
dtype: float64

***Quedarse con solo los registros que tienen un $90$% de sus celdas libres de nulos (nans), luego de este paso deberia quedarte $922772$ registros.***

In [ ]:
prop.dropna(axis = 0, inplace = True, thresh = int(18 * 0.9))

In [ ]:
prop.shape

(922772, 18)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $922772$ filas y $18$ columnas...***

***El analisis se realizará sobre el *AMBA*, eliminar todo lo que no sea parte del mismo. Luego de este paso deberian quedarte $617747$ registros...***

In [ ]:
prop.l2.unique()

array(['Entre Ríos', 'Misiones', 'Santa Fe', 'Bs.As. G.B.A. Zona Oeste',
       'Bs.As. G.B.A. Zona Norte', 'Tucumán',
       'Buenos Aires Costa Atlántica', 'Chubut', 'Capital Federal',
       'Córdoba', 'Buenos Aires Interior', 'Bs.As. G.B.A. Zona Sur',
       'San Luis', 'Neuquén', 'Colonia', 'Río Negro', 'Montevideo',
       'Maldonado', 'Mendoza', 'Corrientes', 'Salta', 'Florida',
       'Tierra Del Fuego', 'Chaco', 'Santa Cruz', 'La Pampa', 'Catamarca',
       'Canelones', 'Santiago Del Estero', 'San Juan', 'Miami', 'Jujuy',
       'Michigan', 'São Paulo', 'La Rioja', 'Rocha', 'Santa Catarina',
       'Pennsylvania', 'Rio Grande do Norte', 'Maryland', 'Formosa',
       'Rio de Janeiro', 'California'], dtype=object)

In [ ]:
prop = prop[(prop.l2 == "Bs.As. G.B.A. Zona Oeste") | 
            (prop.l2 == "Bs.As. G.B.A. Zona Norte") |
            (prop.l2 == "Bs.As. G.B.A. Zona Sur") |
            (prop.l2 == "Capital Federal")]

In [ ]:
prop.shape

(617747, 18)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $617747$ filas y $18$ columnas...***

***Chequear la columna l1 que representa al pais, deberia existir solo un valor: Argentina. Si hay otros valores eliminarlos...***

In [ ]:
prop.l1.unique()

array(['Argentina'], dtype=object)

In [ ]:
prop.l1.value_counts()

Argentina    617747
Name: l1, dtype: int64

***Los registros que tienen el mismo id son de la misma propiedad pero actualizado, quedarse con el mas actualizado, los demas eliminarlos.***

In [ ]:
prop.id.describe()

count                       617747
unique                      617747
top       6TO5dWwmivuJ67zkztUzCA==
freq                             1
Name: id, dtype: object

In [ ]:
prop.id.duplicated().sum()

0

***Eliminar las columnas:***

- **id** porque ya sabemos que todos los registros son diferentes y no vamos a usar esta columna para identificarlos.
- **l1** porque solo tiene un valor: Argentina. No aporta informacion a los datos.
- **description** y **title** porque no contamos con herramientas de NLP para volverlas informacion para visualizar o modelizar.


In [ ]:
prop.columns

Index(['id', 'ad_type', 'start_date', 'end_date', 'created_on', 'lat', 'lon',
       'l1', 'l2', 'l3', 'rooms', 'bathrooms', 'price', 'currency', 'title',
       'description', 'property_type', 'operation_type'],
      dtype='object')

In [ ]:
prop.shape

(617747, 18)

In [ ]:
prop = prop.drop(columns=["id","l1","description","title"])

In [ ]:
prop.columns

Index(['ad_type', 'start_date', 'end_date', 'created_on', 'lat', 'lon', 'l2',
       'l3', 'rooms', 'bathrooms', 'price', 'currency', 'property_type',
       'operation_type'],
      dtype='object')

In [ ]:
prop.shape

(617747, 14)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $617747$ filas y $14$ columnas...***

***Ver los primeros registros del dataframe...***

In [ ]:
prop.head()

ad_type  start_date    end_date  created_on        lat        lon  \
6   Propiedad  2020-08-22  2020-09-04  2020-08-22 -34.604195 -58.559827   
10  Propiedad  2020-08-22  2020-08-25  2020-08-22 -34.439610 -58.629968   
16  Propiedad  2020-08-22  2020-09-04  2020-08-22 -34.633053 -58.419906   
17  Propiedad  2020-08-22  2020-09-04  2020-08-22 -34.397401 -58.649185   
18  Propiedad  2020-08-22  2020-09-04  2020-08-22 -34.397401 -58.649185   

                          l2               l3  rooms  bathrooms    price  \
6   Bs.As. G.B.A. Zona Oeste  Tres de Febrero    2.0        1.0      NaN   
10  Bs.As. G.B.A. Zona Norte            Tigre    4.0        4.0      NaN   
16           Capital Federal            Boedo    NaN        NaN  70000.0   
17  Bs.As. G.B.A. Zona Norte            Tigre    NaN        2.0  35000.0   
18  Bs.As. G.B.A. Zona Norte            Tigre    NaN        2.0  27000.0   

   currency property_type     operation_type  
6       NaN  Departamento           Alquiler  
10      NaN          Casa  Alquiler temporal  
16      ARS       Oficina           Alquiler  
17      ARS       Oficina           Alquiler  
18      ARS       Oficina           Alquiler

***Sospechamos que las columnas ```start_date``` y ```created_on``` en este caso tienen los mismos valores, chequearlo, si son iguales eliminar la columna ```created_on```...***

In [ ]:
prop.columns

Index(['ad_type', 'start_date', 'end_date', 'created_on', 'lat', 'lon', 'l2',
       'l3', 'rooms', 'bathrooms', 'price', 'currency', 'property_type',
       'operation_type'],
      dtype='object')

In [ ]:
prop.shape

(617747, 14)

In [ ]:
(prop.loc[:,'start_date'] == prop.loc[:,'created_on']).sum()

617747

In [ ]:
prop = prop.drop(columns=["created_on"])

In [ ]:
prop.shape

(617747, 13)

***Si hiciste todo bien para este momento deberias tener un dataframe de $(617747, 13)$***

***Ver los primeros registros del dataframe...***

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $617747$ filas y $13$ columnas...***

In [ ]:
prop.head()

ad_type  start_date    end_date        lat        lon  \
6   Propiedad  2020-08-22  2020-09-04 -34.604195 -58.559827   
10  Propiedad  2020-08-22  2020-08-25 -34.439610 -58.629968   
16  Propiedad  2020-08-22  2020-09-04 -34.633053 -58.419906   
17  Propiedad  2020-08-22  2020-09-04 -34.397401 -58.649185   
18  Propiedad  2020-08-22  2020-09-04 -34.397401 -58.649185   

                          l2               l3  rooms  bathrooms    price  \
6   Bs.As. G.B.A. Zona Oeste  Tres de Febrero    2.0        1.0      NaN   
10  Bs.As. G.B.A. Zona Norte            Tigre    4.0        4.0      NaN   
16           Capital Federal            Boedo    NaN        NaN  70000.0   
17  Bs.As. G.B.A. Zona Norte            Tigre    NaN        2.0  35000.0   
18  Bs.As. G.B.A. Zona Norte            Tigre    NaN        2.0  27000.0   

   currency property_type     operation_type  
6       NaN  Departamento           Alquiler  
10      NaN          Casa  Alquiler temporal  
16      ARS       Oficina           Alquiler  
17      ARS       Oficina           Alquiler  
18      ARS       Oficina           Alquiler

***Vamos a analizar la columna ```ad_type```, que tipos de datos unicos tiene, si solo tiene un tipo de dato categorico eliminarla porque no aporta información....***

In [ ]:
prop.shape

(617747, 13)

In [ ]:
prop.ad_type.value_counts()

Propiedad    617747
Name: ad_type, dtype: int64

In [ ]:
prop = prop.drop(columns=["ad_type"])

In [ ]:
prop.shape

(617747, 12)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $617747$ filas y $12$ columnas...***

### AMBA

***Vamos a analizar la columna l3, para eso primero vamos a filtrar solo los lugares que pertenecen al AMBA, los cuales son:***
    
```python
municipios = ['Ciudad Autónoma de Buenos Aires', 'Almirante Brown', 'Avellaneda', 'Berazategui', 'Esteban Echeverría', 'Ezeiza', 'Florencio Varela', 'General San Martín', 'Hurlingham', 'Ituzaingó', 'José C. Paz', 'La Matanza', 'Lanús', 'Lomas de Zamora', 'Malvinas Argentinas', 'Merlo', 'Moreno', 'Morón', 'Quilmes', 'San Fernando', 'San Isidro', 'San Miguel', 'Tigre', 'Tres de Febrero', 'Vicente López']
```

***No vamos a directamente filtrar por los que esten en esta lista, porque no nos fiamos de los datos y tal vez hay alguno que simplemente en el dataset esta escrito de otra forma, por eso primero vamos a ver los registros cuya columna l3 no pertenece a la lista de municipios...***

In [ ]:
municipios = ['Ciudad Autónoma de Buenos Aires', 'Almirante Brown', 'Avellaneda', 'Berazategui', 'Esteban Echeverría', 'Ezeiza', 'Florencio Varela', 'General San Martín', 'Hurlingham', 'Ituzaingó', 'José C. Paz', 'La Matanza', 'Lanús', 'Lomas de Zamora', 'Malvinas Argentinas', 'Merlo', 'Moreno', 'Morón', 'Quilmes', 'San Fernando', 'San Isidro', 'San Miguel', 'Tigre', 'Tres de Febrero', 'Vicente López']

In [ ]:
prop.loc[~(prop.l3.isin(municipios))].l3.unique()

array(['Boedo', 'Recoleta', 'Balvanera', 'San Nicolás',
       'General Rodríguez', 'Almagro', 'Palermo', 'Belgrano', 'La Plata',
       'San Cristobal', 'Barrio Norte', 'Constitución', 'Retiro',
       'Catalinas', 'Monserrat', 'Tribunales', 'Once', 'Caballito',
       'Villa Urquiza', 'Escobar', 'Pilar', 'Nuñez', 'Puerto Madero',
       'Floresta', 'Saavedra', 'Chacarita', 'Villa Crespo',
       'Villa Pueyrredón', 'Flores', 'Coghlan', 'Congreso',
       'Villa Devoto', 'Barracas', 'Colegiales', 'Villa del Parque',
       'San Vicente', 'José C Paz', 'Centro / Microcentro', 'San Telmo',
       'Monte Castro', 'Villa General Mitre', 'Parque Chacabuco', nan,
       'Abasto', 'Marcos Paz', 'Parque Chas', 'Villa Santa Rita',
       'Villa Lugano', 'Liniers', 'Mataderos', 'Las Cañitas', 'Paternal',
       'Villa Luro', 'Parque Patricios', 'Presidente Perón', 'Agronomía',
       'Villa Ortuzar', 'Villa Soldati', 'Parque Centenario', 'Cañuelas',
       'Parque Avellaneda', 'Pompeya', 'Boca'

***Sacar de la seleccion anterior los que pertenecen a Capital Federal y los valores NaN, luego ver cuales son los valores unicos de l3 para ver si queda alguno que pertenece a los municipios pero se escribio con alguna diferencia...***

In [ ]:
capital_federal = list(prop[(prop.l2 == "Capital Federal")].l3.unique())

In [ ]:
prop.loc[~(prop.l3.isin(capital_federal))].l3.unique()

array(['Tres de Febrero', 'Tigre', 'General Rodríguez', 'La Plata',
       'Vicente López', 'Lanús', 'Ituzaingó', 'Ezeiza', 'San Miguel',
       'Lomas de Zamora', 'La Matanza', 'Merlo', 'Escobar', 'Pilar',
       'San Isidro', 'Moreno', 'Morón', 'Avellaneda', 'Almirante Brown',
       'Esteban Echeverría', 'Quilmes', 'Hurlingham', 'San Fernando',
       'General San Martín', 'San Vicente', 'Malvinas Argentinas',
       'José C Paz', 'Berazategui', 'Marcos Paz', 'Florencio Varela',
       'Presidente Perón', 'Cañuelas'], dtype=object)

***Vemos que hay un municipio que se escribio de forma diferente y lo hubiesemos eliminado, teniendo en cuenta esto eliminar todos los registros que no pertenezcan al AMBA, ojo: no eliminar los que tengan valor NaN***

In [ ]:
ambant = ['General Rodríguez',
 'La Plata',
 'Escobar',
 'Pilar',
 'San Vicente',
 'Marcos Paz',
 'Presidente Perón',
 'Cañuelas']

In [ ]:
prop = prop[(prop.l3 != "General Rodríguez") &
            (prop.l3 != "La Plata") &
            (prop.l3 != "Escobar") &
            (prop.l3 != "Pilar") &
            (prop.l3 != "San Vicente") &
            (prop.l3 != "Marcos Paz") &
            (prop.l3 != "Presidente Perón") &
            (prop.l3 != "Cañuelas")]

In [ ]:
prop.shape

(528442, 12)

***Para este paso deberias tener 528442 rows × 12 columns...***

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $528442$ filas y $12$ columnas...***

### **Latitud y longitud**

***Vamos a analizar la calidad de los datos en las columnas de lat y lon. Buscaremos posibles outliers o datos incorrectos, para eso calcula para ambas columnas: el promedio, el maximo y el minimo...***

In [ ]:
round(prop.lat.describe(),2)

count    500380.00
mean        -34.60
std           0.43
min         -53.79
25%         -34.65
50%         -34.60
75%         -34.56
max          42.43
Name: lat, dtype: float64

In [ ]:
round(prop.lon.describe(),2)

count    500341.00
mean        -58.49
std           0.45
min        -122.46
25%         -58.55
50%         -58.46
75%         -58.40
max          58.70
Name: lon, dtype: float64

***Evidentemente hay ciertos puntos que son incorrectos, sus coordenadas caen fuera del AMBA. Para eso buscamos cuales son los valores frontera para la latitud y longitud en el AMBA.***

***Estimamos un segmentos de latitudes de [-35, -34] y longitudes de [-59 a -58]***

***Eliminar todo lo que este fuera de esos rangos (no eliminar valores NaN)***

In [ ]:
prop = prop[prop.lat.between(-35, -34) & prop.lon.between(-59, -58)]

In [ ]:
prop.shape

(499739, 12)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $499739$ filas y $12$ columnas...***

***Volver a chequear estadisticos de lat y lon para ver si ya saque valores incorrectos...***

In [ ]:
round(prop.lat.describe(),2)

count    499739.00
mean        -34.61
std           0.10
min         -34.99
25%         -34.65
50%         -34.60
75%         -34.56
max         -34.05
Name: lat, dtype: float64

In [ ]:
round(prop.lon.describe(),2)

count    499739.00
mean        -58.48
std           0.11
min         -58.98
25%         -58.55
50%         -58.46
75%         -58.40
max         -58.02
Name: lon, dtype: float64

***Eliminar los NaN de la columna l3***

In [ ]:
prop.isna().sum().sort_values(ascending = False)

rooms             96894
bathrooms         76701
currency           5251
price              4694
l3                  106
start_date            0
end_date              0
lat                   0
lon                   0
l2                    0
property_type         0
operation_type        0
dtype: int64

In [ ]:
round((prop.isna().sum()*100/prop.shape[0]),2).sort_values(ascending = False)

rooms             19.39
bathrooms         15.35
currency           1.05
price              0.94
l3                 0.02
start_date         0.00
end_date           0.00
lat                0.00
lon                0.00
l2                 0.00
property_type      0.00
operation_type     0.00
dtype: float64

In [ ]:
prop.dropna(subset=["l3"], inplace = True)

In [ ]:
prop.shape

(499633, 12)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $499633$ filas y $12$ columnas...***

***Volvamos a ver el porcentaje de nulos de cada columna...***

In [ ]:
round((prop.isna().sum()*100/prop.shape[0]),2).sort_values(ascending = False)

rooms             19.39
bathrooms         15.34
currency           1.05
price              0.94
start_date         0.00
end_date           0.00
lat                0.00
lon                0.00
l2                 0.00
l3                 0.00
property_type      0.00
operation_type     0.00
dtype: float64

***Veamos los valores unicos de la columna ```property_type```***

In [ ]:
prop.property_type.unique()

array(['Departamento', 'Casa', 'Oficina', 'Local comercial', 'Lote',
       'Cochera', 'PH', 'Otro', 'Depósito', 'Casa de campo'], dtype=object)

***La columna se ve bien, hacemos lo mismo con ```operation_type```***

In [ ]:
prop.operation_type.unique()

array(['Alquiler', 'Alquiler temporal', 'Venta'], dtype=object)

***La columna se ve bien. Nos quedan price, currency, bathrooms y rooms.***

***Vamos a currency, veamos los valores unicos de esta columna...***

In [ ]:
prop.currency.unique()

array([nan, 'ARS', 'USD', 'COP'], dtype=object)

***Tenemos el peso argentino, el dolar, el peso colombiano y NaN. Veamos los registros con NaN.***

In [ ]:
prop[(prop.currency != "ARS") & (prop.currency != "USD") & (prop.currency != "COP")].currency.unique()

array([nan], dtype=object)

***Vemos que la cantidad de NaN en esta columna es del $1$%, sin esta columna no podremos saber en que moneda esta el precio, por lo que tampoco sabremos el precio (el precio es la columna mas interesante de este dataset). Decidimos eliminar los registros con NaN de la columna currency.***

In [ ]:
prop.dropna(subset=["currency"], inplace = True)

In [ ]:
prop.currency.unique()

array(['ARS', 'USD', 'COP'], dtype=object)

In [ ]:
prop.shape

(494382, 12)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $494382$ filas y $12$ columnas...***

***Volver a analizar el porcentaje de NaN de cada columna.***

In [ ]:
round((prop.isna().sum()*100/prop.shape[0]),2).sort_values(ascending = False)

rooms             19.57
bathrooms         15.49
start_date         0.00
end_date           0.00
lat                0.00
lon                0.00
l2                 0.00
l3                 0.00
price              0.00
currency           0.00
property_type      0.00
operation_type     0.00
dtype: float64

***Vemos que se eliminaron los NaN de price al eliminar los de currency, matamos 2 pajaros de un tiro.***

Ahora si crear una nueva columna **llamada precio** con el valor de las propiedades **en dolares**. Utilizar estas cotizaciones

1 dolar = $283 pesos argentinos

1 dolar = $4637 pesos colombianos

In [ ]:
prop.currency.unique()

array(['ARS', 'USD', 'COP'], dtype=object)

In [ ]:
condicion = [(prop.currency == "ARS"),
             (prop.currency == "COP"),
             (prop.currency == "USD")]

In [ ]:
eleccion = [(prop.price / 283), (prop.price / 4637), (prop.price)]

In [ ]:
prop["precio"] = np.select(condicion, eleccion)

In [ ]:
prop.shape

(494382, 13)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $494382$ filas y $13$ columnas...***

### Alquiler temporal

***Vamos a centrar el analisis en Alquiler y Venta, eliminamos los registros de 'Alquiler temporal'...***

In [ ]:
prop.operation_type.unique()

array(['Alquiler', 'Alquiler temporal', 'Venta'], dtype=object)

In [ ]:
prop = prop[(prop.operation_type != "Alquiler temporal")]

In [ ]:
prop.operation_type.unique()

array(['Alquiler', 'Venta'], dtype=object)

In [ ]:
prop.shape

(472774, 13)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $472774$ filas y $13$ columnas...***

***Ordenar por la columna precio de forma tanto ascendente como descendente y ver los registros mas caros y mas baratos para buscar posibles datos incorrectos...***

In [ ]:
prop.sort_values(by=["precio"], ascending=False)

start_date    end_date        lat        lon  \
419972  2021-01-03  2021-01-04 -34.592273 -58.447200   
339889  2021-01-02  2021-01-02 -34.592273 -58.447200   
128721  2021-01-20  9999-12-31 -34.592273 -58.447200   
339888  2021-01-02  2021-01-02 -34.592273 -58.447200   
979848  2020-10-02  9999-12-31 -34.611003 -58.386067   
...            ...         ...        ...        ...   
541347  2020-11-25  2021-01-14 -34.533896 -58.707248   
941289  2020-11-24  2020-11-25 -34.539896 -58.693022   
555089  2020-10-15  2020-10-30 -34.602049 -58.372480   
998918  2020-12-04  9999-12-31 -34.668022 -58.551872   
364137  2021-04-07  2021-05-28 -34.702419 -58.324239   

                              l2                    l3  rooms  bathrooms  \
419972           Capital Federal             Chacarita    NaN        NaN   
339889           Capital Federal             Chacarita    NaN        NaN   
128721           Capital Federal             Chacarita    NaN        NaN   
339888           Capital Federal             Chacarita    NaN        NaN   
979848           Capital Federal              Congreso    1.0        NaN   
...                          ...                   ...    ...        ...   
541347  Bs.As. G.B.A. Zona Norte            San Miguel    1.0        1.0   
941289  Bs.As. G.B.A. Zona Norte            San Miguel    1.0        1.0   
555089           Capital Federal  Centro / Microcentro    4.0        2.0   
998918  Bs.As. G.B.A. Zona Oeste            La Matanza    1.0        1.0   
364137    Bs.As. G.B.A. Zona Sur            Avellaneda    3.0        1.0   

             price currency property_type operation_type      precio  
419972  35000000.0      USD          Lote          Venta  35000000.0  
339889  35000000.0      USD          Lote          Venta  35000000.0  
128721  35000000.0      USD          Lote          Venta  35000000.0  
339888  35000000.0      USD          Lote          Venta  35000000.0  
979848  26653528.0      USD          Otro          Venta  26653528.0  
...            ...      ...           ...            ...         ...  
541347         0.0      ARS  Departamento       Alquiler         0.0  
941289         0.0      ARS  Departamento       Alquiler         0.0  
555089         0.0      ARS       Oficina       Alquiler         0.0  
998918         0.0      USD  Departamento          Venta         0.0  
364137         0.0      ARS  Departamento       Alquiler         0.0  

[472774 rows x 13 columns]

In [ ]:
prop.sort_values(by=["precio"], ascending=True)

start_date    end_date        lat        lon  \
541349  2020-11-25  2020-12-07 -34.583248 -58.494516   
364137  2021-04-07  2021-05-28 -34.702419 -58.324239   
941289  2020-11-24  2020-11-25 -34.539896 -58.693022   
278063  2020-11-17  2020-11-17 -34.638337 -58.504331   
555089  2020-10-15  2020-10-30 -34.602049 -58.372480   
...            ...         ...        ...        ...   
360288  2020-09-30  2020-10-01 -34.611003 -58.386067   
339889  2021-01-02  2021-01-02 -34.592273 -58.447200   
419972  2021-01-03  2021-01-04 -34.592273 -58.447200   
128721  2021-01-20  9999-12-31 -34.592273 -58.447200   
339888  2021-01-02  2021-01-02 -34.592273 -58.447200   

                              l2                    l3  rooms  bathrooms  \
541349           Capital Federal      Villa Pueyrredón    1.0        1.0   
364137    Bs.As. G.B.A. Zona Sur            Avellaneda    3.0        1.0   
941289  Bs.As. G.B.A. Zona Norte            San Miguel    1.0        1.0   
278063           Capital Federal             Caballito    1.0        1.0   
555089           Capital Federal  Centro / Microcentro    4.0        2.0   
...                          ...                   ...    ...        ...   
360288           Capital Federal              Congreso    1.0        NaN   
339889           Capital Federal             Chacarita    NaN        NaN   
419972           Capital Federal             Chacarita    NaN        NaN   
128721           Capital Federal             Chacarita    NaN        NaN   
339888           Capital Federal             Chacarita    NaN        NaN   

             price currency property_type operation_type      precio  
541349         0.0      ARS  Departamento       Alquiler         0.0  
364137         0.0      ARS  Departamento       Alquiler         0.0  
941289         0.0      ARS  Departamento       Alquiler         0.0  
278063         0.0      USD  Departamento          Venta         0.0  
555089         0.0      ARS       Oficina       Alquiler         0.0  
...            ...      ...           ...            ...         ...  
360288  26653528.0      USD          Otro          Venta  26653528.0  
339889  35000000.0      USD          Lote          Venta  35000000.0  
419972  35000000.0      USD          Lote          Venta  35000000.0  
128721  35000000.0      USD          Lote          Venta  35000000.0  
339888  35000000.0      USD          Lote          Venta  35000000.0  

[472774 rows x 13 columns]

### Registros entre 40 dolares o 5000000

***Eliminar los registros cuyo precio sea inferior a 40 dolares o mayor a 5.000.000 dolares...***

In [ ]:
prop.shape

(472774, 13)

In [ ]:
prop = prop[~((prop.precio <= 40) | (prop.precio >= 5000000))]

In [ ]:
prop.shape

(467968, 13)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $467968$ filas y $13$ columnas...***

***Eliminar las ventas cuyo precio sea menor a 5000 dolares***

In [ ]:
prop.operation_type.unique()

array(['Alquiler', 'Venta'], dtype=object)

In [ ]:
prop = prop.loc[~((prop.operation_type == "Venta") & (prop.precio < 5000)),:]

In [ ]:
prop.shape

(467586, 13)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $467586$ filas y $13$ columnas...***

***Vamos ahora a la columna bathrooms, eliminamos los registros cuyo valor en bathrooms sea mayor al de rooms + 1***

In [ ]:
prop.rooms.unique()

array([nan,  3.,  1.,  2.,  4.,  5.,  7.,  6.,  8., 12.,  9., 10., 15.,
       29., 19., 25., 11., 24., 18., 14., 17., 40., 35., 36., 20., 23.,
       21., 16., 30., 22., 26., 32., 13., 31., 27., 34., 28.])

In [ ]:
prop.bathrooms.unique()

array([nan,  2.,  3.,  1.,  4.,  5.,  8., 10.,  6.,  7., 11.,  9., 16.,
       12., 15., 18., 20., 13., 17., 19., 14.])

In [ ]:
prop = prop[~(prop.bathrooms > (prop.rooms + 1))]

In [ ]:
prop.shape

(465964, 13)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $465964$ filas y $13$ columnas...***

***Ordenar el dataset por la cantidad de baños tanto ascendente como de forma descendente para ver posibles registros incorrectos...***

In [ ]:
prop.sort_values(by=["bathrooms"], ascending=False)

start_date    end_date        lat        lon  \
242841  2020-07-20  9999-12-31 -34.606558 -58.371653   
932251  2020-06-05  2020-06-08 -34.616042 -58.366659   
520542  2020-08-12  2020-08-25 -34.607353 -58.372332   
965124  2021-02-08  9999-12-31 -34.621021 -58.381737   
242864  2020-07-20  9999-12-31 -34.616042 -58.366659   
...            ...         ...        ...        ...   
999598  2020-12-04  9999-12-31 -34.638566 -58.689860   
999673  2020-12-04  2021-07-02 -34.807659 -58.396121   
999819  2020-12-04  2020-12-16 -34.422371 -58.657596   
999928  2020-12-04  9999-12-31 -34.614237 -58.394551   
999938  2020-12-04  2020-12-16 -34.651733 -58.530917   

                              l2                    l3  rooms  bathrooms  \
242841           Capital Federal           San Nicolás    NaN       20.0   
932251           Capital Federal         Puerto Madero    NaN       20.0   
520542           Capital Federal  Centro / Microcentro    NaN       20.0   
965124           Capital Federal          Constitución    NaN       20.0   
242864           Capital Federal         Puerto Madero    NaN       20.0   
...                          ...                   ...    ...        ...   
999598  Bs.As. G.B.A. Zona Oeste             Ituzaingó    4.0        NaN   
999673    Bs.As. G.B.A. Zona Sur       Almirante Brown    5.0        NaN   
999819  Bs.As. G.B.A. Zona Norte                 Tigre    7.0        NaN   
999928           Capital Federal             Balvanera    5.0        NaN   
999938  Bs.As. G.B.A. Zona Oeste       Tres de Febrero    5.0        NaN   

            price currency property_type operation_type     precio  
242841    55000.0      USD       Oficina       Alquiler    55000.0  
932251    73500.0      USD       Oficina       Alquiler    73500.0  
520542    42340.0      USD       Oficina       Alquiler    42340.0  
965124    63470.0      USD       Oficina       Alquiler    63470.0  
242864    91000.0      USD       Oficina       Alquiler    91000.0  
...           ...      ...           ...            ...        ...  
999598   120000.0      USD          Casa          Venta   120000.0  
999673   300000.0      USD          Lote          Venta   300000.0  
999819  1750000.0      USD          Casa          Venta  1750000.0  
999928   235000.0      USD  Departamento          Venta   235000.0  
999938    98000.0      USD  Departamento          Venta    98000.0  

[465964 rows x 13 columns]

In [ ]:
prop.sort_values(by=["bathrooms"], ascending=True)

start_date    end_date        lat        lon  \
512322  2021-05-14  2021-05-17 -34.555853 -58.455217   
452271  2020-09-08  2021-03-17 -34.593986 -58.441347   
452272  2020-09-08  2020-11-10 -34.606274 -58.452012   
452273  2020-09-08  2020-12-03 -34.597223 -58.443872   
452274  2020-09-08  2020-09-22 -34.594978 -58.441647   
...            ...         ...        ...        ...   
999598  2020-12-04  9999-12-31 -34.638566 -58.689860   
999673  2020-12-04  2021-07-02 -34.807659 -58.396121   
999819  2020-12-04  2020-12-16 -34.422371 -58.657596   
999928  2020-12-04  9999-12-31 -34.614237 -58.394551   
999938  2020-12-04  2020-12-16 -34.651733 -58.530917   

                              l2               l3  rooms  bathrooms  \
512322           Capital Federal         Belgrano    2.0        1.0   
452271           Capital Federal     Villa Crespo    1.0        1.0   
452272           Capital Federal     Villa Crespo    2.0        1.0   
452273           Capital Federal     Villa Crespo    2.0        1.0   
452274           Capital Federal     Villa Crespo    2.0        1.0   
...                          ...              ...    ...        ...   
999598  Bs.As. G.B.A. Zona Oeste        Ituzaingó    4.0        NaN   
999673    Bs.As. G.B.A. Zona Sur  Almirante Brown    5.0        NaN   
999819  Bs.As. G.B.A. Zona Norte            Tigre    7.0        NaN   
999928           Capital Federal        Balvanera    5.0        NaN   
999938  Bs.As. G.B.A. Zona Oeste  Tres de Febrero    5.0        NaN   

            price currency property_type operation_type        precio  
512322    39500.0      ARS  Departamento       Alquiler  1.395760e+02  
452271    80000.0      USD  Departamento          Venta  8.000000e+04  
452272   105000.0      USD  Departamento          Venta  1.050000e+05  
452273   115000.0      USD  Departamento          Venta  1.150000e+05  
452274   130000.0      USD  Departamento          Venta  1.300000e+05  
...           ...      ...           ...            ...           ...  
999598   120000.0      USD          Casa          Venta  1.200000e+05  
999673   300000.0      USD          Lote          Venta  3.000000e+05  
999819  1750000.0      USD          Casa          Venta  1.750000e+06  
999928   235000.0      USD  Departamento          Venta  2.350000e+05  
999938    98000.0      USD  Departamento          Venta  9.800000e+04  

[465964 rows x 13 columns]

### Mas de 8 baños o mas de 15 habitaciones

***Eliminar los registros que tengan mas de $8$ baños o mas de $15$ habitaciones...***

In [ ]:
prop.shape

(465964, 13)

In [ ]:
prop = prop.loc[~((prop.bathrooms > 8) | (prop.rooms > 15)),:]

***Eliminar la columna ```price``` y ```currency```, ya que vamos a utilizar la columna precio...***

In [ ]:
prop = prop.drop(columns=["price", "currency"])

In [ ]:
prop.shape

(465451, 11)

**PUNTO DE REVISIÓN**

***A esta altura el dataframe tiene $465451$ filas y $11$ columnas...***

### ***Rellenado***

- ***Rellenar los NaN de bathrooms con la mediana de la columna bathrooms, hacer lo mismo con la columna rooms.***

Pero antes vamos a marcar los registros a los cuales vamos a inferir, para eso **primero de rellenar**, crea una columna llamada **rellenado** la cual tendra un 1 si alguno de sus campos fue rellenado por nosotros o 0 si no fue rellenado.

In [ ]:
prop.bathrooms.unique()

array([nan,  2.,  3.,  1.,  4.,  5.,  8.,  6.,  7.])

In [ ]:
prop.rooms.unique()

array([nan,  3.,  1.,  2.,  4.,  5.,  7.,  6.,  8., 12.,  9., 10., 15.,
       11., 14., 13.])

In [ ]:
cond2 = [(prop['rooms'].isna()), (prop['bathrooms'].isna()), (~(prop['rooms'].isna())), (~(prop['bathrooms'].isna()))]

selec2 = 1, 1, 0, 0

prop['rellenado'] = np.select(cond2, selec2)
prop

start_date    end_date        lat        lon  \
16      2020-08-22  2020-09-04 -34.633053 -58.419906   
17      2020-08-22  2020-09-04 -34.397401 -58.649185   
18      2020-08-22  2020-09-04 -34.397401 -58.649185   
19      2020-08-22  2020-09-04 -34.598112 -58.400784   
20      2020-08-22  2020-09-04 -34.604054 -58.393770   
...            ...         ...        ...        ...   
999990  2020-12-04  2020-12-16 -34.462664 -58.625235   
999991  2020-12-04  2021-01-05 -34.473210 -58.760048   
999993  2020-12-04  9999-12-31 -34.482248 -58.501989   
999996  2020-12-04  9999-12-31 -34.596587 -58.376893   
999997  2020-12-04  9999-12-31 -34.576445 -58.432603   

                              l2                   l3  rooms  bathrooms  \
16               Capital Federal                Boedo    NaN        NaN   
17      Bs.As. G.B.A. Zona Norte                Tigre    NaN        2.0   
18      Bs.As. G.B.A. Zona Norte                Tigre    NaN        2.0   
19               Capital Federal             Recoleta    NaN        3.0   
20               Capital Federal            Balvanera    NaN        1.0   
...                          ...                  ...    ...        ...   
999990  Bs.As. G.B.A. Zona Norte                Tigre    NaN        5.0   
999991  Bs.As. G.B.A. Zona Norte  Malvinas Argentinas    NaN        2.0   
999993  Bs.As. G.B.A. Zona Norte           San Isidro   11.0        4.0   
999996           Capital Federal               Retiro   12.0        5.0   
999997           Capital Federal              Palermo   14.0        2.0   

          property_type operation_type        precio  rellenado  
16              Oficina       Alquiler  2.473498e+02          1  
17              Oficina       Alquiler  1.236749e+02          1  
18              Oficina       Alquiler  9.540636e+01          1  
19              Oficina       Alquiler  2.296820e+02          1  
20              Oficina       Alquiler  2.120141e+02          1  
...                 ...            ...           ...        ...  
999990             Casa          Venta  1.200000e+06          1  
999991             Casa          Venta  1.120000e+05          1  
999993             Casa          Venta  6.900000e+05          0  
999996  Local comercial       Alquiler  3.700000e+03          0  
999997  Local comercial          Venta  5.700000e+05          0  

[465451 rows x 12 columns]

***Calcular la cantidad de registros rellenados y no rellenados...***

In [ ]:
prop.rellenado.value_counts()

0    345611
1    119840
Name: rellenado, dtype: int64

***Ahora si: Rellená los NaN de bathrooms con la mediana de la columna bathrooms, hacer lo mismo con la columna rooms con la mediana de la columna rooms...***

In [ ]:
media_bathrooms = prop.bathrooms.mean()

In [ ]:
media_rooms = prop.rooms.mean()

In [ ]:
prop.bathrooms.fillna(value=media_bathrooms, inplace = True)
prop.rooms.fillna(value=media_rooms, inplace = True)

***Volver a analizar el porcentaje de NaN de cada columna. No deberia de existir ningun NaN en nuestro dataframe...***

In [ ]:
round((prop.isna().sum()*100/prop.shape[0]),2).sort_values(ascending = False)

start_date        0.0
end_date          0.0
lat               0.0
lon               0.0
l2                0.0
l3                0.0
rooms             0.0
bathrooms         0.0
property_type     0.0
operation_type    0.0
precio            0.0
rellenado         0.0
dtype: float64

In [ ]:
prop.shape

(465451, 12)

***Parece ser que el formato se encuentra correcto, vamos a aplicar una pequeña comprobacion logica: la fecha start_date puede ser igual o menor que end_date, nunca mayor. Si ocurre esto, eliminar los registros que contienen estos errores.***

In [ ]:
prop[(prop.start_date <= prop.end_date)]

start_date    end_date        lat        lon  \
16      2020-08-22  2020-09-04 -34.633053 -58.419906   
17      2020-08-22  2020-09-04 -34.397401 -58.649185   
18      2020-08-22  2020-09-04 -34.397401 -58.649185   
19      2020-08-22  2020-09-04 -34.598112 -58.400784   
20      2020-08-22  2020-09-04 -34.604054 -58.393770   
...            ...         ...        ...        ...   
999990  2020-12-04  2020-12-16 -34.462664 -58.625235   
999991  2020-12-04  2021-01-05 -34.473210 -58.760048   
999993  2020-12-04  9999-12-31 -34.482248 -58.501989   
999996  2020-12-04  9999-12-31 -34.596587 -58.376893   
999997  2020-12-04  9999-12-31 -34.576445 -58.432603   

                              l2                   l3      rooms  bathrooms  \
16               Capital Federal                Boedo   2.798034   1.642851   
17      Bs.As. G.B.A. Zona Norte                Tigre   2.798034   2.000000   
18      Bs.As. G.B.A. Zona Norte                Tigre   2.798034   2.000000   
19               Capital Federal             Recoleta   2.798034   3.000000   
20               Capital Federal            Balvanera   2.798034   1.000000   
...                          ...                  ...        ...        ...   
999990  Bs.As. G.B.A. Zona Norte                Tigre   2.798034   5.000000   
999991  Bs.As. G.B.A. Zona Norte  Malvinas Argentinas   2.798034   2.000000   
999993  Bs.As. G.B.A. Zona Norte           San Isidro  11.000000   4.000000   
999996           Capital Federal               Retiro  12.000000   5.000000   
999997           Capital Federal              Palermo  14.000000   2.000000   

          property_type operation_type        precio  rellenado  
16              Oficina       Alquiler  2.473498e+02          1  
17              Oficina       Alquiler  1.236749e+02          1  
18              Oficina       Alquiler  9.540636e+01          1  
19              Oficina       Alquiler  2.296820e+02          1  
20              Oficina       Alquiler  2.120141e+02          1  
...                 ...            ...           ...        ...  
999990             Casa          Venta  1.200000e+06          1  
999991             Casa          Venta  1.120000e+05          1  
999993             Casa          Venta  6.900000e+05          0  
999996  Local comercial       Alquiler  3.700000e+03          0  
999997  Local comercial          Venta  5.700000e+05          0  

[465451 rows x 12 columns]

***Vamos a pasar las columnas start_date y end_date, que son strings, a formato datetime. Recorda que el formato de fecha que tenemos en nuestras columnas de fecha es %Y/%m/%d***

In [ ]:
prop.start_date = pd.to_datetime(prop["start_date"], format = "%Y/%m/%d")

In [ ]:
#prop.end_date = pd.to_datetime(prop["end_date"], format = "%Y/%m/%d")

***Al tratar de pasar la columna end_date a datetime nos sale un error, hay un registro que lo esta causando, investigar que registro causa este error...***

***Hay muchos registros que tienen como end_date 9999-12-31, como la documentacion es de mala calidad, la misma no especifica que significa que el end_date tenga este valor, pero aplicando un poco de sentido comun deducimos que los registros que tienen esta fecha significa que el aviso no se dio de baja, por lo cual no se vendio.***

***Cambiar los registros con esta fecha, cambiandola a la fecha de la primera publicacion de python por parte de Guido van Rossum:***
**1991-02-20**. 

***Los registros con esta fecha seran los que aun no se vendieron.***

In [ ]:
prop['end_date'].replace('9999-12-31', '1991-02-20', inplace=True)

In [ ]:
prop.shape

(465451, 12)

***Ahora si pasar la columna end_date a datetime...***

In [ ]:
prop.end_date = pd.to_datetime(prop["end_date"], format = "%Y/%m/%d")

In [ ]:
prop.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 465451 entries, 16 to 999997
Data columns (total 12 columns):
 #   Column          Non-Null Count   Dtype         
---  ------          --------------   -----         
 0   start_date      465451 non-null  datetime64[ns]
 1   end_date        465451 non-null  datetime64[ns]
 2   lat             465451 non-null  float64       
 3   lon             465451 non-null  float64       
 4   l2              465451 non-null  object        
 5   l3              465451 non-null  object        
 6   rooms           465451 non-null  float64       
 7   bathrooms       465451 non-null  float64       
 8   property_type   465451 non-null  object        
 9   operation_type  465451 non-null  object        
 10  precio          465451 non-null  float64       
 11  rellenado       465451 non-null  int64         
dtypes: datetime64[ns](2), float64(5), int64(1), object(4)
memory usage: 46.2+ MB


***Vamos a cambiar el nombre de las columnas de la siguiente manera:...***

- start_date => inicio
- end_date => fin
- lat => lat
- lon => lon
- l2 => zona
- l3 => municipio
- rooms => ambientes
- bathrooms => baños
- property_type => tipo
- operation_type => operacion
- precio => precio
- rellenado => rellenado

In [ ]:
columnas = ['Inicio',
           'Fin',
           'Lat',
           'Lon',
           'Zona',
           'Municipio',
           'Ambientes',
           'Baños',
           'Tipo',
           'Operacion',
           'Precio',
           'Rellenado']

In [ ]:
prop.columns = columnas

In [ ]:
prop.head(3)

Inicio        Fin        Lat        Lon                      Zona  \
16 2020-08-22 2020-09-04 -34.633053 -58.419906           Capital Federal   
17 2020-08-22 2020-09-04 -34.397401 -58.649185  Bs.As. G.B.A. Zona Norte   
18 2020-08-22 2020-09-04 -34.397401 -58.649185  Bs.As. G.B.A. Zona Norte   

   Municipio  Ambientes     Baños     Tipo Operacion      Precio  Rellenado  
16     Boedo   2.798034  1.642851  Oficina  Alquiler  247.349823          1  
17     Tigre   2.798034  2.000000  Oficina  Alquiler  123.674912          1  
18     Tigre   2.798034  2.000000  Oficina  Alquiler   95.406360          1

***Crear dos columnas nuevas:***

- ***año_inicio:*** tiene que contener el numero de año que hay en la fecha de la columna inicio
- ***mes_inicio:*** tiene que contener el numero de mes que hay en la fecha de la columna inicio

In [ ]:
prop['Año_inicio'] = pd.DatetimeIndex(prop['Inicio']).year

prop['Mes_inicio'] = pd.DatetimeIndex(prop['Inicio']).month

***Por ultimo, reinicia el indice...***

In [ ]:
prop.reset_index(drop=True, inplace=True)
prop.head(5)

Inicio        Fin        Lat        Lon                      Zona  \
0 2020-08-22 2020-09-04 -34.633053 -58.419906           Capital Federal   
1 2020-08-22 2020-09-04 -34.397401 -58.649185  Bs.As. G.B.A. Zona Norte   
2 2020-08-22 2020-09-04 -34.397401 -58.649185  Bs.As. G.B.A. Zona Norte   
3 2020-08-22 2020-09-04 -34.598112 -58.400784           Capital Federal   
4 2020-08-22 2020-09-04 -34.604054 -58.393770           Capital Federal   

   Municipio  Ambientes     Baños     Tipo Operacion      Precio  Rellenado  \
0      Boedo   2.798034  1.642851  Oficina  Alquiler  247.349823          1   
1      Tigre   2.798034  2.000000  Oficina  Alquiler  123.674912          1   
2      Tigre   2.798034  2.000000  Oficina  Alquiler   95.406360          1   
3   Recoleta   2.798034  3.000000  Oficina  Alquiler  229.681979          1   
4  Balvanera   2.798034  1.000000  Oficina  Alquiler  212.014134          1   

   Año_inicio  Mes_inicio  
0        2020           8  
1        2020           8  
2        2020           8  
3        2020           8  
4        2020           8

In [ ]:
prop.shape

(465451, 14)

### ***Comprobacion***

***Si hasta aca lo hiciste bien te deberia quedar un dataset de $465451$ rows $×$ $14$ columns y que tenga los siguientes parametros iguales:...***



In [ ]:
if (465451 - prop.shape[0]) == 0:
  print(f"Vamo nomaaaaaaaaaa")
else:
  print(f"Segui participando, todavia tenes una diferencia de {465451 - prop.shape[0]}")

Vamo nomaaaaaaaaaa


***Guardar el dataframe con el nombre 'propiedades_limpio.csv'***

## **Análisis de datos**

***El dueño de la inmobiliaria donde pertencemos no sabe lo que quiere, tampoco entiende que datos interesantes podriamos sacar de este dataset.***

***Por eso nuestro objetivo es encontrar posibles datos curiosos sobre el mismo.*** 

***Algunos ejemplos de datos curiosos podrian ser:***

- cantidad de propiedades por zona
- media de precios por zona
- media de precios por municipio
- promedio de precio por tipo de venta
- cantidad de ventas y de alquileres
- Graficos de barras o de linea, precio vs ambientes
- Cualquier cosa que te parezca interesante


